parquet/src/main/java/org/apache/iceberg/parquet/ParquetFileMerger.java

```diff
@@ -0,0 +1,684 @@
+   * @param fileIO FileIO to use for reading files
+   * @param targetOutputSize Maximum size for output file (files larger than this cannot be merged)
+   * @return MessageType schema if files can be merged, null otherwise
+   */
+  public static MessageType canMergeAndGetSchema(
+      List<DataFile> dataFiles, FileIO fileIO, long targetOutputSize) {
+    if (dataFiles == null || dataFiles.isEmpty()) {
+      return null;
+    }
+
+    // Single loop to check partition spec consistency, file sizes, and build InputFile list
+    int firstSpecId = dataFiles.get(0).specId();
+    List<InputFile> inputFiles = Lists.newArrayListWithCapacity(dataFiles.size());
+    for (DataFile dataFile : dataFiles) {
+      // Check partition spec consistency - all files must have the same spec
+      if (dataFile.specId() != firstSpecId) {
+        return null;
+      }
+
+      // Check file sizes - don't merge if splitting large files
+      if (dataFile.fileSizeInBytes() > targetOutputSize) {
+        return null;
+      }
+
+      inputFiles.add(fileIO.newInputFile(dataFile.path().toString()));
+    }
+
+    return canMergeAndGetSchema(inputFiles);
+  }
+
+  /**
+   * Reads the Parquet schema from an Iceberg InputFile.
+   *
+   * @param inputFile Iceberg input file to read schema from
+   * @return MessageType schema of the Parquet file
+   * @throws IOException if reading fails
+   */
+  private static MessageType readSchema(InputFile inputFile) throws IOException {
+    return ParquetFileReader.open(ParquetIO.file(inputFile))
+        .getFooter()
+        .getFileMetaData()
+        .getSchema();
+  }
+
+  /**
+   * Validates that all row lineage column values are non-null in the input files.
+   *
+   * <p>When files already have physical row lineage columns and we're doing row lineage processing,
+   * we cannot automatically calculate null values during binary merge. This method ensures all
+   * values in both _row_id and _last_updated_sequence_number columns are present.
+   *
+   * @param inputFiles List of input files to validate
+   * @throws IllegalArgumentException if any row lineage column contains null values
+   * @throws IOException if reading file metadata fails
+   */
+  private static void validateRowLineageColumnsHaveNoNulls(List<InputFile> inputFiles)
+      throws IOException {
+    for (InputFile inputFile : inputFiles) {
+      try (ParquetFileReader reader = ParquetFileReader.open(ParquetIO.file(inputFile))) {
+        List<BlockMetaData> rowGroups = reader.getFooter().getBlocks();
+
+        for (BlockMetaData rowGroup : rowGroups) {
+          for (ColumnChunkMetaData columnChunk : rowGroup.getColumns()) {
+            String columnPath = columnChunk.getPath().toDotString();
+
+            // Check if this is the _row_id column
+            if (columnPath.equals(MetadataColumns.ROW_ID.name())) {
+              Statistics<?> stats = columnChunk.getStatistics();
+              if (stats != null && stats.getNumNulls() > 0) {
```

Comment:
Is it possible that the column is there, but the stats are not calculated? Could we rewrite the rowIds by mistake in this case?

Reply:
Good catch!  Changed it to 
```
if (stats == null || stats.getNumNulls() > 0) {
 return false;
 }
```